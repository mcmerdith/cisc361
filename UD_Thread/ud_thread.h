/*
 * thread library function prototypes
 */

// Create a new thread from FUNCTION with id THREAD_ID and priority PRIORITY
void t_create(void (*function)(int), int thread_id, int priority);

// Terminate a thread. Semantically, this should be called at the end of every thread function
void t_terminate(void);

// Voluntarily release CPU control and be placed at the end of the thread queue
void t_yield(void);

// Initialize the thread library
void t_init(void);

// Shutdown and cleanup the thread library
void t_shutdown(void);
