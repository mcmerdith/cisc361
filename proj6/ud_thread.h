/*
 * thread library function prototypes
 */

// Create a new thread from FUNCTION with id THREAD_ID and priority PRIORITY
void t_create(void (*function)(int), int thread_id, int priority);

// Terminate a thread. Semantically, this should be called at the end of every thread function
void t_terminate(void);

// Voluntarily release CPU control and be placed at the end of the thread queue
void t_yield(void);

// Initialize the thread library
void t_init(void);

// Shutdown and cleanup the thread library
void t_shutdown(void);

/**
 * semaphore
 */

// Create a new semaphore pointed to by sp with a count value of sem_count.
int sem_init(sem_t **sp, int sem_count);

// Current thread does a wait (P) on the specified semaphore.
void sem_wait(sem_t *sp);

// Current thread does a signal (V) on the specified semaphore.
// Follow the Mesa semantics (p. 9 of OSTEP Chapter 30 Condition Variables) where the thread that signals continues,
// and the first waiting (blocked) thread (if there is any) becomes ready.
void sem_signal(sem_t *sp);

// Free any memory related to the specified semaphore.
// If there are queued TCBs, they are moved to the ready queue, and this may imply that the semantics of the application might be incorrect.
void sem_destroy(sem_t **sp);

typedef void sem_t;