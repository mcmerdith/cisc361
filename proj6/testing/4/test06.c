/*
 * Test Program #6 - Mailbox
 */

#include <stdio.h>
#include <stdlib.h>
#include "ud_thread.h"

mbox *mb;
char *msg[2] = {"hello world...", "bye, bye"};

void producer(int id)
{
    int i;
    char mymsg[30];

    for (i = 0; i < 2; i++)
    {
        sprintf(mymsg, "%s - tid %d", msg[i], id);
        printf("Producer (%d): [%s] [length=%d]\n", id, mymsg, (int)strlen(mymsg));
        mbox_deposit(mb, mymsg, strlen(mymsg));
    }

    t_terminate();
}

void consumer(int id)
{
    int i;
    int len;
    char mesg[1024];

    for (i = 0; i < 4; i++)
    {
        mbox_withdraw(mb, mesg, &len);
        printf("Message from mailbox: [%s]\n", mesg);
    }

    t_terminate();
}

int main(void)
{

    t_init();

    mbox_create(&mb);
    t_create(producer, 1, 1);
    t_create(producer, 2, 1);
    t_create(consumer, 3, 1);
    t_yield();
    mbox_destroy(&mb);

    t_shutdown();
    printf("Done with mailbox test...\n");

    return 0;
}