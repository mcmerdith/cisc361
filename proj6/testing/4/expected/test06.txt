Producer (1): [hello world... - tid 1] [length=22]
Producer (1): [bye, bye - tid 1] [length=16]
Producer (2): [hello world... - tid 2] [length=22]
Producer (2): [bye, bye - tid 2] [length=16]
Message from mailbox: [hello world... - tid 1]
Message from mailbox: [bye, bye - tid 1]
Message from mailbox: [hello world... - tid 2]
Message from mailbox: [bye, bye - tid 2]
Done with mailbox test...
